### ISSUE TITLE
In 8. regex replace is 1.5x-2.5x slower than in older versions (7.x, 6.x)

### ISSUE THREAD

**Issue by felix-22** [notepad-plus-plus/notepad-plus-plus#10398]
Related issue: https://github.com/notepad-plus-plus/notepad-plus-plus/issues/10301#issuecomment-895237160

### Description of the Issue

Regex replace in N++ 8.x is drastically slower than in older versions of N++ (7.x, 6.x)
Latest good version is 7.9.5

### Steps to Reproduce the Issue

1. Open [THIS FILE](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8279#issuecomment-743696790)
2. Replace what: `^.+NotepadPP.+\r\n`
3. Replace with: nothing

### Expected Behavior

It should not take 1.5x-2.5x more time than in older versions (7.x, 6.x)

### Actual Behavior

It takes 1.5x-2.5x more time than on older versions of N++ (7.x, 6.x)
Test repeated on various 8.x versions (32bit and 64bit) gives results in sec: 93 / 79 / 66 / 56 / 53 / 47 what is ~1.5x - ~2,5x slower, it seems the time spent is often random, for ex. one time it might spent 1.5x time, another time it can spent 2.5x time on the same file, like depending on the humor...

### Debug Information

Test done on default install, no additional plugins installed. Win 7 SP1 64
Not sure whether it's a regression in Notepad or in Scintilla.
Also, as regex replace is one of the most used and needed functions, the issue should get high priority.

EDIT: full debug:
```
Notepad++ v8.1.3   (32-bit)
Build time : Aug 10 2021 - 00:32:53
Path : X:\xxxxxx_xxxx_2\xxxxxxx_xxxxxxxx\xxxxxxxx.xxxxx_xxxxxxx_xxxxxxx\_xxxxxxx\xxxxxxxx_xxxx\npp.8.1.3.portable\notepad++.exe
Command Line : 
Admin mode : OFF
Local Conf mode : ON
Cloud Config : OFF
OS Name : Windows 7 Professional (64-bit) 
OS Build : 7601.0
Current ANSI codepage : 1250
Plugins : mimeTools.dll NppConverter.dll NppExport.dll 
```

**Comment by mere-human** [comment]
> Debug Information

@felix-22  No, you should select a menu `? > Debug info > Copy debug info into clipboard` then past it here.

**Comment by felix-22** [comment]
Okay then, I've updated the initial comment.

**Comment by mere-human** [comment]
I think I saw a similar problem before. Could you, please, try doing the same steps but in a 64-bit version of app?
You can download a portable version just to try it so that your current 32-bit installation is intact.

**Comment by felix-22** [comment]
I already did, like I said in the initial comment: "Test repeated on various 8.x versions (32bit and 64bit)", but at your request I just have repeated test on the 64bit ver. and it still made no difference.

However I've found something different: the replace regex speed is a little incerased (16%) when running as admin compared to when not running as admin, but since it applies to both the old 7.x and the new 8.x, it does not make any difference to the issue as the gap between 7.x and 8.x remains the same. Example results for the latest test: 

```
ver    | as admin | not as admin
7.9.5     30 sec       36 sec
8.1.3     47 sec       53 sec
```

**Comment by mere-human** [comment]
@chcg @Uhf7 Any ideas? Could it be related to Boost.Regex update?

**Comment by chcg** [comment]
Likely this is related to the boost.regex change https://github.com/notepad-plus-plus/notepad-plus-plus/pull/9594 , mainly relaxed by https://github.com/notepad-plus-plus/notepad-plus-plus/pull/10010.
Ideas:
- remove further locale stuff from https://github.com/notepad-plus-plus/notepad-plus-plus/pull/9707
- build with Profiling to check where the time is spent
- Maybe a check of C++11 regex vs boost.regex gives some hint, if it is just boost is responsible for the slowdown.

**Comment by donho** [comment]
> Add a performance check target to the appveyor build to find performance regressions.

That's the hardest part - Even using a private machine for this purpose, the result of the performance test cannot be constant. AppVeyor's machine is not dedicated for Notepad++ project so their CPU usage could vary largely.

**Comment by donho** [comment]
@chcg It seems this perf issue is due to #9707 
After remove all locale stuff I can see the good perf is back.
Thank you for pointing it out.

**Comment by donho** [comment]
@felix-22 
Could you try [fixed x32](download.notepad-plus-plus.org/repository/MISC/npp_fix_perf_32.zip) or [x64  binary](download.notepad-plus-plus.org/repository/MISC/npp_fix_perf_64.zip) ?
I believe this issue is fixed in the binaries I provide above.

**Comment by felix-22** [comment]
Yeah, it's fixed:

```
ver   | as admin | not as admin
32bit     29 s        35 s
64bit     23 s        30 s
```

By they way, any idea why it is 17%-23% faster when running as admin?

**Comment by mere-human** [comment]
> After remove all locale stuff I can see the good perf is back.

It seems we would have to come up with a different way to handle locale-specific text in regex.

**Comment by donho** [comment]
@felix-22 
Thank you for confirmation and heads up for the link problem.

> By they way, any idea why it is 17%-23% slower when not running as admin?

I have absolutely no idea about it. I suppose it's rather a system issue for accessing some files while user has no admin privilege.

**Comment by donho** [comment]
@mere-human 

> It seems we would have to come up with a different way to handle locale-specific text in regex.

Yes, absolutely. There should be some boost parameter for fixing this issue (instead of using Win32 routine).

**Comment by Uhf7** [comment]
Just investigated the issue. It seems to me, that not the Boost regex library is slowing down the process, but the changing of the locale itself. It seems, that the ```std::locale::global(std::locale("")``` call and it's restore after each search is the reason for the slow-down. It is more expensive than I thought. But setting the locale to "" fixes the original problems. So I would try to provide a solution, in which the locale is set to "" before "Find all" or "Replace all", and restored after it. Then, the good influence of the locale "" to the original issues could be kept, and it should be as fast as before.

**Comment by mere-human** [comment]
Why don't we set the locale once, when we start the app (or change the language)?
Why can't we use the user's preferred native locale always?

**Comment by Uhf7** [comment]
The reason why I have chosen the set/restore approach was, that I saw usage of locales in some other places in the software, and I didn't want to disturb their behavior by changing anything globally. 

Setting the locale once at start of Notepad++ sounds good for me. But there is an actually "unsolvable" task before it: We need to investigate all API calls, which are influenced by the current locale, and this is a quite hard thing to do. That's why I took the set/restore approach.

But it is daring: Set the locale to "" at start of Notepad++, and see what get's broken. I would support this.

The second best, and quite save approach, is to save the current locale before all "Find all" and "Replace all" invocations and  restore it after.

On the other hand: Set the locale once at start, I like this much more. How do we make sure, that nothing gets broken?
